### ISSUE TITLE
Do this drivers support the "Partial" messages

### ISSUE THREAD

**Issue by danielmotaleite** [MickayG/moby-kafka-logdriver#3]
As per https://github.com/moby/moby/pull/32947 , docker now breaks the logs in 16k blocks and flag then with "Partial" bool field. 
Not having tried this yet nor know enough about the docker plugin code, does this driver plugin support this incomplete messages and rebuild then?

If not, can you add support for this?

Thanks

**Comment by MickayG** [comment]
@danielmotaleite  Not at present but I can add support for that soon

**Comment by MickayG** [comment]
@danielmotaleite I read through the ticket you referenced and looked at the journald docker plugin code (as you mentioned it supported it). It may not be so easy with this plugin. 

We shouldn't reconstruct the partial messages on the write to Kafka as we could consume far too much memory and negate the reason for the splitting in the first place.  Journald does not reconstruct when storing, only when retrieving the log messages from the store and playing back to the user.

The second issue is that there is no guarantee of correct ordering when consuming from multiple Kafka partitions, so reconstructing partial messages where those messages are in different partitions would be problematic. 

It would be possible to do the reconstruction as long as all the parts of the incomplete message were in a single partition from using a 'partition by key' partitioning strategy and "key by container id" but that's a very specific case.   I'll do some further reading and come up with an idea

**Comment by danielmotaleite** [comment]
That is true, but systemd journald cheats, it's storage drive is linear, so it is easy for then to reconstruct the event later...

This need to be merged as soon as possible, because later you start to get with out of order, different partitions or servers, different consumers and similar problems... it is the logstash multiline support problem all again (that after several years trying to fix it, it was dropped in favor of a simple multiline input and recommendation to merge the lines as soon as possible, before logstash)

So i would say that the docker plugin is the correct place, is the first step after the docker broke the message in 16k blocks. the logger plugin can then merge back the messages in the correct order from the correct container and to avoid using all the ram, one can have a configurable limit.

With this limit if one have 1MB event, set the plugin to 1.2MB. Those that have 18k events, can set the limit to 20k. If someones sets the limit to 16GB, well, it is their own fault if they get out of RAM :)

i think that the configurable limit should exist in docker, but as they denied it and pushed it to the plugins, then the plugins are the ones that should solve the docker broken events.
